SCClassLibrary/External/Ableton/SCClockMeterSync.sc

```diff
@@ -85,10 +85,16 @@ SCClockMeterSync {
 		} { broadcastMeter = saveFlag };
 	}
 
+	// must always call action after timeout!
+	// resyncMeter depends on this
 	queryMeter { |action, timeout = 0.2|
-		var replies = IdentityDictionary.new,
-		resp;
+		var replies, resp;
 		if(clock.numPeers > 0) {
+			if(timeout.isKindOf(SimpleNumber).not or: { timeout <= 0 }) {
```

Comment:
what would be passed instead of a number?

Reply:
This is per a review comment from Brian: It's critical that this method not wait forever, but it's possible to pass in (erroneously) an object that will cause it to hang indefinitely. It's not a bad thing, I think, to prevent that before it happens.

Reply:
But wouldn't it be better to throw an error as usual?

Reply:
Do we have an established practice of what to do if one thread is hanging for an asynchronous result, but the asynchronous operation fails with an exception?

I don't think we do. That will be the case here.

So, maybe I should add an `onError` argument to queryMeter.

Reply:
I think yes, or maybe just within `resyncMeter`, where it is used, have a protect:

```supercollider
protect {
	this.queryMeter { |val|
	replies = val;
	cond.unhang;
	}
} {
	cond.unhang;
}
```

Reply:
The place where the error would occur is forked onto a separate thread. queryMeter has already returned, protect has already returned, you can't catch the error this way.

It's the same problem that I was trying to explain on the mailing list about Scott C.'s Promise-like class -- you can't throw an error directly back into the thread that was waiting. I ended up dealing with it by saving the error in a collection associated with the thread -- then, resuming the thread went back into the Promise's wait method, where we check for an error belonging to the thread and throw at that time.

But now we are getting out of scope of this PR, as we seem to love to do. This is actually kind of a tricky issue and I'd suggest that perhaps an already-large PR is not the place to establish a new best practice.

Reply:
>you can't throw an error directly back into the thread that was waiting.

I assumed that was the case, so no digression on my part. OK, we need to investigate into this. I thought that it does this (see `handleError`), but now have forgotten how I made it  work back then.

Reply:
> queryMeter has already returned
I don't see why it would return if an error is thrown in it.

But if you all think this is a bigger or different issue, forget about my suggestion.

Reply:
> I don't see why it would return if an error is thrown in it.

The error is not thrown in the method!

It's thrown in a routine that happened to have been forked within the method.

Anywhere else in the codebase, if someone wrote `myMethod { fork { ... } }`, it would be understood that the sequence of events is:

1. The method schedules a routine on a clock.

2. The method returns `this`.

3. When the interpreter is not busy, the scheduled routine awakes.

The error will be thrown in stage 3. *Not* 1 or 2.

It's too late to bounce the error back to the caller because the caller is already forgotten (stack frames already traced back).

Reply:
My suggestion here is to mark the location in the codebase with some comment that we can grep later, and open an issue for the broader question of catching errors thrown in forked routines (i.e., we are in a forked routine, where a Condition or such is `hang`ing, and the thing it's waiting for needs to throw an error back), then circle back and update here.

I can imagine how to do it by adding something to Condition, but then this PR would become responsible not only for a new class, documentation and unit tests directly related to the issue at hand, but also for new methods, documentation and unit tests for something not really related... let alone, as Brian said on the mailing list, sync issues are notoriously difficult to get right. I think in this case (perhaps not every case), the design considerations should not be tacked onto an unrelated changeset.

Reply:
yes, but there are details. let's keep the details for another discussion!

Reply:
Oh, wait, I see what you mean... I was thinking of Brian's earlier comment that the "wait" expression would throw (which is inside the fork) but we could just throw from this `if`, before forking anything. (After literally 6-7 hours of addressing review comments, my mind was apparently a bit jelly.)

It probably is better to throw from here. Will do next time I'm at the computer.

Reply:
yes, that was actually what I meant. Thank you for taking a look!

Reply:
Done.